socketio/src/asynchronous/client/client.rs

```diff
@@ -203,9 +204,31 @@ impl Client {
             }
         });
 
+        self.poll_future = Some(
+            async {
+                if let Err(err) = poll_handle.await {
+                    error!("Polling task returned an error: {}", err)
+                }
+            }
+            .boxed()
+            .shared(),
+        );
+
         Ok(())
     }
 
+    /// Waits until the stream started by [`Client::poll_stream`] terminates,
+    /// at which point the socket.io connection to the server is shut down.
+    ///
+    /// If the connection has not been started yet, this terminates immediately,
+    /// so make sure to call [`Client::poll_stream`] (done by [`ClientBuilder::connect`])
+    /// first.
+    pub async fn wait_for_completion(&self) {
```

Comment:
I think this should return some kind of result (like an error reason?), as it can fail in various ways:

* Client was never connected
* Client did successfully disconnect itself (graceful shutdown requested).
* Server sent an error or whatever.

Reply:
> Client did successfully disconnect itself (graceful shutdown requested).

We could make the `DisconnectReason` public and return it here.

> Server sent an error or whatever.

That's handled through `.on(Event::Error`.

> Client was never connected

We could return a Result which contains an error if the client was not connected and the disconnect reason otherwise.

Reply:
Then why don't handle the server-side disconnect via the `on` callbacks and make this poll function blocking? I see some kind of split API here and I feel like not knowing which method to use when.